src/commands/install.ts

```diff
@@ -30,43 +42,92 @@ export default class Install extends Command {
+      ServiceConfig.saveToPath(root_service_path, config);
+      cli.action.stop(chalk.green(`${args.service_name} installed`));
+    } else {
+      await this.installServices((root_service as LocalServiceNode), dependency_graph, flags.recursive);
     }
+  }
+
+  async get_remote_definitions(remote_service_version: string) {
+    const repository_name = url.resolve(`${this.app.config.registry_host}/`, `${remote_service_version}`);
 
-    if (!config.dependencies) {
-      config.dependencies = {};
+    let config;
+    try {
+      config = await this.load_service_config(repository_name);
+    } catch {
+      await execa('docker', ['pull', repository_name]);
+      config = await this.load_service_config(repository_name);
     }
+    return config;
+  }
 
-    config.dependencies[service_name] = service_tag;
-    ServiceConfig.saveToPath(service_dir, config);
-    this.log(chalk.green(`${service_name}:${service_tag} installed successfully`));
+  async load_service_config(repository_name: string) {
+    const { stdout } = await execa('docker', ['inspect', repository_name, '--format', '{{ index .Config.Labels "architect.json"}}']);
+    const config = JSON.parse(stdout);
+    if (config.api) {
+      const { stdout } = await execa('docker', ['inspect', repository_name, '--format', '{{ index .Config.Labels "api_definitions"}}']);
+      return JSON.parse(stdout);
+    }
   }
 
-  async run() {
-    const {args} = this.parse(Install);
+  async installServices(root_dependency: LocalServiceNode, all_dependencies: DependencyGraph, recursive: boolean) {
+    const service_dependencies = [root_dependency];
+    const _seen = [];
+    while (service_dependencies.length) {
+      const target_dependency = service_dependencies.pop();
+      _seen.push(target_dependency!.name);
 
-    if (args.service_name) {
-      await this.installSingle(args.service_name);
-    } else {
-      // TODO: codegen for existing services
+      cli.action.start(chalk.blue(`Installing ${target_dependency!.name}`), undefined, { stdout: true });
+      await ProtocExecutor.execute(target_dependency!, (target_dependency as LocalServiceNode));
+      cli.action.stop(chalk.green(`${target_dependency!.name} installed`));
+
+      const directDependencies = all_dependencies.getNodeDependencies(target_dependency!);
+      for (const dependency of directDependencies) {
+        if (!(dependency instanceof DatastoreNode)) {
+          cli.action.start(chalk.blue(`Installing ${dependency.name} as dependency of ${target_dependency!.name}`), undefined, { stdout: true });
+          await ProtocExecutor.execute(target_dependency!, (dependency as LocalServiceNode));
+          cli.action.stop(chalk.green(`${dependency.name} installed`));
+          if (recursive && !_seen.includes(dependency.name)) {
+            service_dependencies.push(dependency as LocalServiceNode);
+          }
+        }
+      }
```

Comment:
This would be much cleaner as a fixed loop across all graph nodes:

```js
async installServices(graph: DependencyGraph) {
  graph.nodes.forEach((node, key) => {
    const dependencies = graph.getNodeDependencies(node);
    for (const dependency of dependencies) {
      if (!(dependency instanceof DatastoreNode)) {
        cli.action.start(chalk.blue(`Installing ${dependency.name} as dependency of ${target_dependency!.name}`), undefined, { stdout: true });
        await ProtocExecutor.execute(node!, (dependency as LocalServiceNode));
        cli.action.stop(chalk.green(`${dependency.name} installed`));
      }
    }
  });
}
```

Reply:
Do you still want this changed after our conversation on the way to lunch?

Reply:
Yes please